File: apps/daemon/graite/api/voice.py
"""Realtime voice with the assistant: readiness, the duplex audio socket, and its voice.

The socket is the daemon's second WebSocket (after `/events`). Client -> server: binary
PCM16LE 16 kHz mono microphone audio in any chunking, and JSON controls
(`interrupt`, `played`, `mute`, `barge_in`, `end`). Server -> client: JSON events and binary
PCM16LE audio at the rate announced in `ready`, belonging to the last `audio_start`.
"""

from __future__ import annotations

import asyncio
import io
import json
import logging
import tempfile
import wave
from pathlib import Path
from typing import Any

from fastapi import APIRouter, HTTPException, Request, Response, WebSocket, WebSocketDisconnect
from pydantic import BaseModel, Field

from graite.api.auth import websocket_authorized
from graite.voice.runtime import Busy
from graite.voice.tts import Voice
from graite.voice.voices import BUILT_IN, VoiceInfo

log = logging.getLogger("graite.voice")
router = APIRouter(tags=["Voice"])
UI_ACTOR = "ui"
REFERENCE_MIN_SECONDS = 3
REFERENCE_MAX_SECONDS = 30
REFERENCE_MAX_BYTES = 40 * 1024 * 1024
CLOSE_UNAUTHORIZED = 4401
CLOSE_BUSY = 4409
CLOSE_NOT_READY = 4412


class Missing(BaseModel):
    key: str
    text: str


class VoiceStatus(BaseModel):
    ready: bool
    missing: list[Missing] = Field(default_factory=list)
    active: bool = False


@router.get("/voice/status", response_model=VoiceStatus)
async def voice_status(request: Request) -> VoiceStatus:
    return VoiceStatus(**request.app.state.voice.status())


@router.websocket("/voice/session")
async def voice_session(websocket: WebSocket) -> None:
    state = websocket.app.state
    if not websocket_authorized(websocket, state.settings.token):
        await websocket.close(code=CLOSE_UNAUTHORIZED)
        return
    await websocket.accept()
    conversation_id = websocket.query_params.get("conversation_id") or ""
    barge_in = websocket.query_params.get("barge_in") != "false"  # on unless switched off
    testing = websocket.query_params.get("mode") == "test"  # the test bench: no assistant
    lock = asyncio.Lock()  # events and audio come from two tasks; frames must not interleave

    async def send_json(data: dict[str, Any]) -> None:
        async with lock:
            await websocket.send_text(json.dumps(data, ensure_ascii=False))

    async def send_bytes(data: bytes) -> None:
        async with lock:
            await websocket.send_bytes(data)

    try:
        if testing:
            session = await state.voice.open_test(send_json, send_bytes)
        else:
            session = await state.voice.open(
                conversation_id, send_json, send_bytes, barge_in=barge_in
            )
    except Busy as exc:
        await send_json({"type": "error", "text": str(exc)})
        await websocket.close(code=CLOSE_BUSY)
        return
    except WebSocketDisconnect:
        return
    except (ValueError, OSError) as exc:
        await send_json({"type": "error", "text": str(exc)})
        await websocket.close(code=CLOSE_NOT_READY)
        return
    try:
        while True:
            message = await websocket.receive()
            if message["type"] == "websocket.disconnect":
                break
            if message.get("bytes") is not None:
                await session.audio(message["bytes"])
            elif message.get("text"):
                try:
                    control = json.loads(message["text"])
                except ValueError:
                    continue
                if not isinstance(control, dict):
                    continue
                if control.get("type") == "end":
                    break
                await session.control(control)
    except WebSocketDisconnect:
        pass
    finally:
        await state.voice.close()
        try:
            await websocket.close()
        except RuntimeError:
            pass


# ----------------------------------------------------------------- voices


class VoiceList(BaseModel):
    voices: list[VoiceInfo]


def _library(request: Request) -> Any:
    return request.app.state.voice.voices


@router.get("/voice/voices", response_model=VoiceList)
async def list_voices(request: Request) -> VoiceList:
    library = _library(request)
    return VoiceList(voices=await asyncio.to_thread(library.all))


@router.post("/voice/voices", response_model=VoiceInfo, status_code=201)
async def add_voice(
    request: Request, name: str, consent: bool = False, filename: str = "voice.wav"
) -> VoiceInfo:
    """A voice from a short recording of one person (the request body). It is converted to
    the mono 24 kHz WAV the voice engine reads, and only stored with the confirmation that
    the user may use this voice."""
    from graite.media.decode import audio_wav

    if not consent:
        raise HTTPException(
            400,
            "Confirm that this is your own voice, or that you have the speaker's permission.",
        )
    data = await request.body()
    if not data or len(data) > REFERENCE_MAX_BYTES:
        raise HTTPException(400, "Use a recording of up to 40 MB.")
    suffix = Path(filename).suffix.lower() or ".wav"

    def prepare() -> tuple[bytes, float]:
        with tempfile.NamedTemporaryFile(suffix=suffix) as handle:
            handle.write(data)
            handle.flush()
            wav = audio_wav(Path(handle.name), rate=24000, max_seconds=REFERENCE_MAX_SECONDS)
        with wave.open(io.BytesIO(wav)) as parsed:
            return wav, parsed.getnframes() / parsed.getframerate()

    try:
        wav, seconds = await asyncio.to_thread(prepare)
    except ValueError as exc:
        raise HTTPException(400, str(exc)) from exc
    except Exception as exc:  # noqa: BLE001 - undecodable input
        raise HTTPException(400, "This recording could not be read.") from exc
    if seconds < REFERENCE_MIN_SECONDS:
        raise HTTPException(400, "Use at least three seconds of clear speech; ten is better.")
    library = _library(request)
    created: VoiceInfo = await asyncio.to_thread(library.add, name, wav, seconds)
    return created


class RenameVoice(BaseModel):
    name: str = Field(min_length=1, max_length=60)


@router.patch("/voice/voices/{voice_id}", response_model=VoiceInfo)
async def rename_voice(voice_id: str, body: RenameVoice, request: Request) -> VoiceInfo:
    try:
        renamed: VoiceInfo = await asyncio.to_thread(_library(request).rename, voice_id, body.name)
    except ValueError as exc:
        raise HTTPException(404, str(exc)) from exc
    return renamed


@router.delete("/voice/voices/{voice_id}")
async def delete_voice(voice_id: str, request: Request) -> dict[str, bool]:
    try:
        await asyncio.to_thread(_library(request).remove, voice_id)
    except ValueError as exc:
        raise HTTPException(404, str(exc)) from exc
    return {"ok": True}


class PreviewBody(BaseModel):
    text: str = Field(default="", max_length=300)
    language: str | None = None
    # A voice from the library to try; omitted = the assistant's current voice.
    voice_id: str | None = None
    exaggeration: float | None = Field(default=None, ge=0, le=2)
    cfg: float | None = Field(default=None, ge=0, le=1)


@router.post("/voice/preview")
async def preview(body: PreviewBody, request: Request) -> Response:
    """One sentence in a voice, as a WAV. The engine stays loaded for a minute afterwards."""
    from graite.models.config import load_config

    state = request.app.state
    runtime = state.voice
    if runtime.session is not None:
        raise HTTPException(409, "Wait for the running conversation to finish.")
    definition = state.definitions.assistant()
    voice, _ = runtime.voice_for(definition)
    reference = voice.reference
    if body.voice_id is not None:
        reference = runtime.voices.path(body.voice_id)
        if reference is None and body.voice_id != BUILT_IN:
            raise HTTPException(404, "This voice no longer exists.")
    language = body.language if body.language and body.language != "auto" else voice.language
    voice = Voice(
        language,
        reference,
        voice.exaggeration if body.exaggeration is None else body.exaggeration,
        voice.cfg if body.cfg is None else body.cfg,
    )
    name = definition.name if definition else "your assistant"
    text = body.text.strip() or f"Hi, I am {name}. This is how I sound."
    output = io.BytesIO()
    try:
        engine = await runtime.sample_engine(load_config(state.db))
        with wave.open(output, "wb") as wav:
            wav.setnchannels(1)
            wav.setsampwidth(2)
            wav.setframerate(engine.sample_rate)
            async for chunk in engine.synthesize(text, voice):
                wav.writeframesraw(chunk)
    except ValueError as exc:
        raise HTTPException(409, str(exc)) from exc
    return Response(output.getvalue(), media_type="audio/wav")

File: apps/desktop/src/lib/voice.ts
import { reportError } from "./clientLog";
import { request } from "./api";
import { platform } from "./platform";
import type { components } from "@graite/api-types";
import type { VoiceServerEvent } from "@graite/api-types";

export type VoiceStatus = components["schemas"]["VoiceStatus"];
export type VoiceInfo = components["schemas"]["VoiceInfo"];
export const BUILT_IN_VOICE = "built-in";

export interface SampleOptions {
  text?: string;
  language?: string;
  /** A voice from the library to try; omitted = the assistant's current voice. */
  voiceId?: string;
  exaggeration?: number;
  cfg?: number;
}

async function binary(path: string, init: RequestInit, failure: string): Promise<Response> {
  const { url, token } = await platform.getDaemonInfo();
  const response = await fetch(`${url}${path}`, {
    ...init,
    headers: { Authorization: `Bearer ${token}`, ...(init.headers ?? {}) },
  });
  if (!response.ok) {
    const error = (await response.json().catch(() => ({}))) as { detail?: string };
    throw new Error(error.detail || `${failure} (${response.status})`);
  }
  return response;
}
export type { VoiceServerEvent };

export const voice = {
  status: () => request<VoiceStatus>("/api/v1/voice/status"),
  /** One sentence in a voice, as a WAV blob, with how long it took to make. */
  sample: async (options: SampleOptions = {}): Promise<{ blob: Blob; seconds: number }> => {
    const started = performance.now();
    const response = await binary(
      "/api/v1/voice/preview",
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          text: options.text ?? "",
          language: options.language,
          voice_id: options.voiceId,
          exaggeration: options.exaggeration,
          cfg: options.cfg,
        }),
      },
      "The sample failed",
    );
    const blob = await response.blob();
    return { blob, seconds: (performance.now() - started) / 1000 };
  },
  voices: async () => (await request<{ voices: VoiceInfo[] }>("/api/v1/voice/voices")).voices,
  addVoice: async (file: Blob, name: string, filename: string, consent: boolean) => {
    const query = new URLSearchParams({ name, filename, consent: String(consent) });
    const response = await binary(
      `/api/v1/voice/voices?${query}`,
      { method: "POST", headers: { "Content-Type": "application/octet-stream" }, body: file },
      "The voice could not be added",
    );
    return (await response.json()) as VoiceInfo;
  },
  renameVoice: (id: string, name: string) =>
    request<VoiceInfo>(`/api/v1/voice/voices/${encodeURIComponent(id)}`, {
      method: "PATCH",
      body: JSON.stringify({ name }),
    }),
  deleteVoice: (id: string) =>
    request<{ ok: boolean }>(`/api/v1/voice/voices/${encodeURIComponent(id)}`, {
      method: "DELETE",
    }),
};

/** Seconds of audio in a 16-bit mono WAV blob made by the daemon (44-byte header). */
export function wavSeconds(blob: Blob, sampleRate = 24000): number {
  return Math.max(0, blob.size - 44) / 2 / sampleRate;
}

export interface VoiceSocketHandlers {
  onEvent: (event: VoiceServerEvent) => void;
  onAudio: (data: ArrayBuffer) => void;
  onClose: (code: number) => void;
}

/** The duplex voice socket: microphone frames up, events and the assistant's audio down. */
export class VoiceSocket {
  private socket: WebSocket | null = null;

  /** `conversationId === null` opens the test bench: microphone, turn-taking and Whisper
   * only, no assistant behind it. */
  async open(
    conversationId: string | null,
    bargeIn: boolean,
    handlers: VoiceSocketHandlers,
  ): Promise<void> {
    const { url, token } = await platform.getDaemonInfo();
    const query = new URLSearchParams(
      conversationId === null
        ? { token, mode: "test" }
        : { token, conversation_id: conversationId, barge_in: String(bargeIn) },
    );
    const socket = new WebSocket(`${url.replace(/^http/, "ws")}/api/v1/voice/session?${query}`);
    socket.binaryType = "arraybuffer";
    socket.onmessage = (message) => {
      try {
        if (typeof message.data === "string")
          handlers.onEvent(JSON.parse(message.data) as VoiceServerEvent);
        else handlers.onAudio(message.data as ArrayBuffer);
      } catch (error) {
        // One bad event or audio chunk must not take the conversation (or the page) down.
        reportError("voice event failed", error, "voice");
      }
    };
    // A failed connection is followed by `onclose`, which ends the call; nothing else to do.
    socket.onerror = () => {};
    socket.onclose = (event) => handlers.onClose(event.code);
    this.socket = socket;
  }

  sendAudio(frame: Int16Array): void {
    if (this.socket?.readyState === WebSocket.OPEN) this.socket.send(frame.buffer as ArrayBuffer);
  }

  send(control: Record<string, unknown>): void {
    if (this.socket?.readyState === WebSocket.OPEN) this.socket.send(JSON.stringify(control));
  }

  close(): void {
    this.send({ type: "end" });
    this.socket?.close();
    this.socket = null;
  }
}
